### PR TITLE
New Resource: aws_guardduty_ipset

### PR DESCRIPTION
**Pull request by atsushi-ishibashi**
#2489 
```
TF_ACC=1 go test ./aws -v -run=TestAccAwsGuardDutyIpset_basic -timeout 120m
=== RUN   TestAccAwsGuardDutyIpset_basic
--- PASS: TestAccAwsGuardDutyIpset_basic (164.15s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	164.185s
```

### CODE REVIEW

File: aws/resource_aws_guardduty_ipset.go [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+		Timeout:    5 * time.Minute,
+		MinTimeout: 3 * time.Second,
+	}
+
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("[WARN] Error waiting for GuardDuty IpSet status to be \"%s\": %s", guardduty.IpSetStatusActive, err)
+	}
+
+	d.SetId(*resp.IpSetId)
```
Comment (bflad) [inline aws/resource_aws_guardduty_ipset.go]: Can you make the resource ID `fmt.Sprintf("%s:%s", detectorID, *resp.IpSetId)` so we can also add the passthrough import please?

File: aws/resource_aws_guardduty_ipset_test.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+  name = "tf-%s"
+  activate = true
+  detector_id = "${aws_guardduty_detector.test.id}"
+  format = "TXT"
+  location = "https://s3.amazonaws.com/${aws_s3_bucket_object.test.bucket}/${aws_s3_bucket_object.test.key}"
+}
+`, testAccGuardDutyDetectorConfig_basic1, rName, rName, rName)
+}
+
+func testAccGuardDutyIpsetConfig_update(rName, modName string) string {
```
Comment (bflad) [inline aws/resource_aws_guardduty_ipset_test.go]: Nitpick: Can you make the `Config_basic` able to take in all the required inputs so we don't need two configuration functions please?

File: aws/validators.go [hunk 3]
```diff
@@ -2133,3 +2133,15 @@ func validateGameliftOperatingSystem(v interface{}, k string) (ws []string, erro
 	}
 	return
 }
+
+func validateGuardDutyIpsetFormat(v interface{}, k string) (ws []string, errors []error) {
+	value := v.(string)
+	validType := []string{"TXT", "STIX", "OTX_CSV", "ALIEN_VAULT", "PROOF_POINT", "FIRE_EYE"}
```
Comment (bflad) [inline aws/validators.go]: Nitpick: Can you use the SDK provided constants please? e.g. `guardduty.IpSetFormatAlienVault`

File: website/docs/r/guardduty_ipset.html.markdown [hunk 4]
```diff
@@ -0,0 +1,53 @@
+---
+layout: aws
+page_title: 'AWS: aws_guardduty_ipset'
+sidebar_current: docs-aws-resource-guardduty-ipset
+description: Provides a resource to manage a GuardDuty IPSet
+---
+
+# aws_guardduty_ipset
+
+Provides a resource to manage a GuardDuty detector.
```
Comment (bflad) [inline website/docs/r/guardduty_ipset.html.markdown]: Copypasta: Should be `GuardDuty IPSet`

File: aws/resource_aws_guardduty_ipset_test.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+	return nil
+}
+
+func testAccCheckAwsGuardDutyIpsetExists(name string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		_, ok := s.RootModule().Resources[name]
+		if !ok {
+			return fmt.Errorf("Not found: %s", name)
+		}
+
```
Comment (bflad) [inline aws/resource_aws_guardduty_ipset_test.go]: Generally we try to verify the resources actually exist in AWS in these. Once the ID is updated, you can do that here. 👍

File: aws/resource_aws_guardduty_ipset.go [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+	resp, err := conn.GetIPSet(input)
+	if err != nil {
+		if isAWSErr(err, guardduty.ErrCodeBadRequestException, "The request is rejected because the input detectorId is not owned by the current account.") {
+			log.Printf("[WARN] GuardDuty IpSet %q not found, removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
```
Comment (bflad) [inline aws/resource_aws_guardduty_ipset.go]: We need to `d.Set("activate", *resp.Status == guardduty.IpSetStatusActive)` here. We may want to consider renaming the attribute `active`.

File: website/docs/r/guardduty_ipset.html.markdown (hunk 4 above)
Comment (bflad) [reply website/docs/r/guardduty_ipset.html.markdown]: We might want to include the note from the [GuardDuty API documentation](https://docs.aws.amazon.com/guardduty/latest/ug/create-ip-set.html) here.
```
~> **Note:** Currently in GuardDuty, users from member accounts cannot upload and further manage IPSets. IPSets that are uploaded by the master account are imposed on GuardDuty functionality in its member accounts.
```

**Comment by atsushi-ishibashi** [comment]
```
TF_ACC=1 go test ./aws -v -run=TestAccAwsGuardDutyIpset_* -timeout 120m
=== RUN   TestAccAwsGuardDutyIpset_basic
--- PASS: TestAccAwsGuardDutyIpset_basic (212.81s)
=== RUN   TestAccAwsGuardDutyIpset_import
--- FAIL: TestAccAwsGuardDutyIpset_import (114.30s)
	testing.go:513: Step 1 error: ImportStateVerify attributes not equivalent. Difference is shown below. Top is actual, bottom is expected.
		
		(map[string]string) {
		}
		
		
		(map[string]string) (len=1) {
		 (string) (len=11) "detector_id": (string) (len=32) "c8b0a11cff4d9f541ec460da3967be55"
		}
		
FAIL
exit status 1
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	327.160s
make: *** [testacc] Error 1
```
I didn't understand how `Importer` works so could you help me? @bflad

**Comment by bflad** [comment]
@atsushi-ishibashi `d.Set("detector_id", detectorId)` in the read function from the value you get from parsing the ID 😄

**Comment by atsushi-ishibashi** [comment]
@bflad Thank you!
```
TF_ACC=1 go test ./aws -v -run=TestAccAwsGuardDutyIpset_import -timeout 120m
=== RUN   TestAccAwsGuardDutyIpset_import
--- PASS: TestAccAwsGuardDutyIpset_import (79.31s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	79.360s
```

**Comment by bflad** [comment]
Thanks so much @atsushi-ishibashi! This looks great and I'll be merging it momentarily. :shipit: 

```
make testacc TEST=./aws TESTARGS='-run=TestAccAwsGuardDutyIpset'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAwsGuardDutyIpset -timeout 120m
=== RUN   TestAccAwsGuardDutyIpset_basic
--- PASS: TestAccAwsGuardDutyIpset_basic (56.88s)
=== RUN   TestAccAwsGuardDutyIpset_import
--- PASS: TestAccAwsGuardDutyIpset_import (35.16s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	92.090s
```

FYI I will be adding one very minor commit to this, its needed to ensure the acceptance testing works when run in parallel testing environments (like our daily repository acceptance testing) since GuardDuty only allows one detector in each account at a time (check out: https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_guardduty_test.go). We don't expect most folks to know or worry about this most of the time since we're very appreciative of `make testacc` runners already! 🥇

**Comment by atsushi-ishibashi** [comment]
@bflad I got it. Thank you!

**Comment by bflad** [comment]
This has been released in [terraform-provider-aws version 1.9.0](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#190-february-09-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Pull request merged**